File: Binary Search/codeforces_1201C.cpp

// Problem link ===>> https://codeforces.com/contest/1201/problem/C
// submission link ===>> https://codeforces.com/contest/1201/submission/312776638

#include <bits/stdc++.h>
using namespace std;
#define ll long long
#define endl '\n'

vector<ll> arr;
ll n, k;
ll minOpeartions(ll m)
{
    ll totalOp = 0;
    for (int i = n / 2; i < n; i++)
    {
        if (arr[i] < m)
            totalOp += m - arr[i];
    }
    return totalOp;
};

int isOk(ll id)
{
    if (minOpeartions(id) > k)
        return 1;
    return 0;
};

int main()
{
    ios_base::sync_with_stdio(false);
    cout.tie(nullptr);
    cin.tie(nullptr);

    cin >> n >> k;
    for (int i = 0, x; i < n; i++)
    {
        cin >> x;
        arr.push_back(x);
    }
    sort(arr.begin(), arr.end());

    ll l = 0, r = 1e9 + 1 + k;
    while (l < r)
    {
        ll mid = (l + r) / 2;
        if (isOk(mid) == 0)
            l = mid + 1;
        else
            r = mid;
    }
    cout << l - 1 << endl;

    return 0;
}

File: Binary Search/codeforces_1201C.py

# Problem link ===>> https://codeforces.com/contest/1201/problem/C

import sys


def min_operations(m, arr, n):
    total_op = 0
    mid_index = n // 2 
    for i in range(mid_index, n):
        if arr[i] < m:
            total_op += m - arr[i]
    return total_op


def is_ok(mid, arr, n, k):
    return min_operations(mid, arr, n) > k


def main():
    n, k = map(int, sys.stdin.readline().split())
    arr = list(map(int, sys.stdin.readline().split()))

    arr.sort()

    left, right = 0, int(1e9 + 1 + k)

    while left < right:
        mid = (left + right) // 2
        if not is_ok(mid, arr, n, k):
            left = mid + 1
        else:
            right = mid

    print(left - 1)

if __name__ == "__main__":
    main()
